Release 4.1.0 | Aesthetic, functionality changes and a rewrite

vladfrangu opened the pull request:
## Things to consider

This version changes **everything**. You will have to update your values from `config.json` and `auth.json` to their respective JS counterparts.

You will also notice 1 pre-assigned API key in `auth.js`, and thats no mistake. 

## Warning

Do **NOT** Run this branch. Don't. At the time of writing this (or rewriting), the bot has **NO** core features available. Its a full rewrite. Wait till I'll edit this and tell you you *may* go ahead and give it a test run, at your own risk.
And, if you find any bugs, or you'd like a code refactor, drop a comment.

## The aim of this Pull Request

This Pull Request is aimed to **rewrite** and improve the bot. Everything will look better, should work faster, and best of all, everything should and will work.

## What does this Pull Request Contain

- A full bot rewrite, to Discord.js

- Improvements to the command looks and feel

- Addition of smaller things

- A new god damn Extension API

## Current Progress

- [x] Delete the old files

- [x] Rewrite the main base, using `async`

- [ ] Port all of the commands

- [x] Fix broken shit (`giveaway`, `lottery`, `MOTD constantly repeating` etc)

- [ ] Test things

- [x] Pray to Vlag

## Old Credit should go where its due

(This is here for consistency, these credits affect the old Eris commits)

Thanks TTtie ( @TTtie ) for the help provided with the `GetUserWithoutREST`, helping me test the `info` command due to me refactoring some code and with the inspiration for the `debug` command

KingDGrizzle commented:
Mark my words @GilbertGobbels if you don't do a full check on all the code and commands I'll kill you!

TTtie commented:
~~and i raprot u to dismcord for telling u hurt gilbert~~

KingDGrizzle commented:
*cough* ![image](https://hypesquad.dabbit-pri.me/d2129c.png)

TTtie commented:
O kurva.....

KingDGrizzle commented:
Hey @GilbertGobbels promise me that you will respect the eslint rules or I'll personally setup travis and tests for eslint which will prevent PR's from merging unless rules are met:tm:

~Don't do a javascript await meme i swear~

KingDGrizzle commented:
Wasn't eslint, i just did it cause

> It works

> It makes more sense

KingDGrizzle commented:
Need to rewrite the main message.. fuck

TTtie commented on Discord.js:
```diff
@@ -0,0 +1,316 @@
+/* eslint-disable max-len */
+module.exports = (db, configJS, configJSON) => {
+	// Create a Discord.js Bot Client
+	const Discord = require("discord.js");
```
rip eris in gab 2016(?) - 2017.

KingDGrizzle commented:
10/10 that was either a sarcastic thanks, or a legit thanks 🤣

KingDGrizzle commented:
When codacity will bang us 😂

TTtie commented:
![DAhaa](https://cdn.discordapp.com/attachments/325211556277256192/341981252338974732/image.png)

KingDGrizzle commented:
Version 4.1 it is! 🎉

TTtie commented:
:eyes: 

![screenshot_20170803-124100](https://user-images.githubusercontent.com/17570452/28918300-0cad72ec-7849-11e7-8c6d-5290a1bbe4dd.png)

KingDGrizzle commented:
Mark this, as both Travis and Codacy gave their checks!

TTtie commented:
👌

KingDGrizzle commented:
So many freakin checks!

TTtie commented:
god bless

vitalkanev commented on Web/views/pages/landing.ejs:
```diff
@@ -607,13 +518,13 @@
 									<div class="content">
-										I do embeds <span class="icon">
-											<i class="fa fa-id-card-u"></i>
+										Command Wizzard <span class="icon">
```
typo here

vladfrangu replied:
Its `Wizard` isn't it? 😄

vitalkanev replied:
yes it is...

oskikiboy commented:
Using figlet module for customisable ascii art

KingDGrizzle commented:
Except we probably don't want that customized so uh.. @GilbertGobbels opinions?

vitalkanev commented:
Can you just fix Travis...? I'm pretty confused

KingDGrizzle commented:
Its the database being down, and thats @GilbertGobbels DB so.. It will get fixed _soon_

TTtie commented:
:tada: TRAVIS WORKS :eyes:

vitalkanev commented:
@TTTie I think there is no "bot run" - only eslint

jtsshieh commented:
Isn't Eris faster than discord.js? Eris doesn't have all the stupid bloaty functions (tag) that discord.js has.

KingDGrizzle commented:
I'm genuinely confused.. Unless you can explain why that is, I'm remaining confused

jtsshieh commented:
Discord.js has a bunch of functions to **Make your life easier** but not necessarily making the bot faster. When aiming for speed, usually Eris is the winner because it doesn't have all the functions that **Make your life easier** but still it maintains the basic functions and is very optimized for speed. So, your title is kinda misleading.

KingDGrizzle commented:
How the? What?!

Just cause it has more functions, it doesn't mean its `slower`.
Thanks to Discord.js, we get a lot of features that are **needed** for our use internally, as well as allows us to easily multi-thread this bot.

Just adding more features to `make your life easier` doesn't mean its more `bloated`. Just use what you want.

TTtie commented:
define the features

Commit:
[Utils/ObjectDefines] Remove these old things

Commit:
[Structures] Add the new Structure expander
Which allows us to insert custom properties in *some* Discord.js classes

Commit:
[Utils] Forgot this

Commit:
[Gist] Stop censoring the user ID, its public data >.>

Commit:
[Events/Message] Adapt to the new message structure

Commit:
Merge branch 'indev-4.0.2' of https://github.com/GilbertGobbels/GAwesomeBot into indev-4.0.2

Commit:
[Discord.js] Some caching issues fixed

Commit:
[Events] Leave blocked guilds at startup or when the bots added.

Commit:
[QoL] Bot = Client

Commit:
[Events/Message] Change the response for the chatter API erroring

Commit:
[SpamHandler] Run a forEach only if there are any messages

Commit:
[Config.json] Bring some comments to config.json as well as add more features

Commit:
[QoL] config.js gets our Discord invite link
Because you get support, he gets support, everyone gets support!

Commit:
[Web] /dashboard/global-options/bot-user

Commit:
[Config] Some fixes hehe

Commit:
[Web/Database/bot/Discord] Traffic stats!
The start of something beautiful. Still missing:
A place in the maintainer dashboard
Some uniqueUser fixes

Coming tomorrow!

Commit:
[Web] Maintainer Overview page overhaul!
Now including a fully working set of Traffic graphs hoo hoo woo woo

Commit:
[Web] /dashboard/global-options/homepage

Commit:
[Traffic] Fixed Global issues
Because I loove complaining about smallest things!

Commit:
Merge branch 'indev-4.0.2' of https://github.com/GilbertGobbels/GAwesomeBot into indev-4.0.2

Commit:
[Deps]

Commit:
Update README.md

Commit:
[*] Update README

Commit:
[Events/Message] Remove suffix from command arguments as its available in the messae object

Commit:
[PMCommands/AFK] First command :Tada:

Commit:
[Commands/8ball] Finished 8ball

Commit:
[Base Files] Updated base files for commands

Commit:
[Events/Message*] Fixed Gilberts change in events

Commit:
[Readme]

Commit:
Update README.md

Not sure what Vlad tried to do here...

Commit:
[*] Add "Fastboot"
A module by devsnek that allows for even faster startup times.
Just make sure to delete package-cache.json before updating modules

Commit:
[Web] /dashboard/global-options/wiki-contributors
We're cranking these out now!

Commit:
[Web] /dashboard/commands/command-options improvements!
You can now disable chatterbot usage in certain channels! :tada:

But, you will HAVE TO run the migrate script and let it do its thing! Otherwise, baad things can happen.

Commit:
[Errors] Remove errors from the global list

Commit:
[Discord.js] Fix docstring typo and simplify if check
(Thanks webstorm ._.)

Commit:
[Modules/Timeouts] Add startedAt date for further use in other places

Commit:
[Web] Spacing woops!

Commit:
[Modules] Improvements to migration script

Commit:
[Web] Tweaks

Commit:
[Misc] Make Webstorm stop complaining

Commit:
[Discord.js] Slight tweaks

Commit:
[*] More extensions, more work, still got a few things to adapt

Commit:
[Web] Lotsa fixes

Commit:
[Boot] Boot module!

Commit:
[Boot] String update

Commit:
[PMCommands/Reload] Remove userDocument from params

Commit:
[Cache] Modularize the cache system some more, in case we need to re-use it somewhere else

Commit:
[Structures] Don't initialize the user document on each user creation

Commit:
[Discord.js] Did a booboo

Commit:
[*] Fix some bugs

Commit:
[PMCommands] Updated base file

Commit:
[PMCommands/config] Finish config command

Commit:
[PMCommand/Reload] Some more tweaks

Commit:
[PMCommands] Fix reload yet again

Commit:
[*Commands] Forgot to add config command here

Commit:
[PMCommand/config] Patch

Commit:
[PMCommands] help and join are now done

Commit:
[Web] /dashboard/maintainer/version [1/2]
it's the first commit in a series of two, I'll be working on the actual updating system now :)

Commit:
[Web] Routing update

Commit:
[PMCommand|*] Part 1 of the profile command is done

Commit:
[PMCommand/profile] slight tweak

Commit:
[Web/Updater] Updater, fully functioning!

Commit:
[Web] Annoy Gilbert but change this string

Commit:
[Web] /dashboard/management/maintainers

Commit:
[Web] get your own updater scrub

Commit:
[Web] String Update

Commit:
[Web] Sudo Maintainer System Basics

Commit:
[Web] The menu now reflects your Maintainer perms

Commit:
[Web] Displaying Additional Perms instead of disabling them
to non-sudo-maintainers, so they can *know* what perms they have, but not modify them.

Commit:
[*] Fix Travis and Update CODEOWNERS

Commit:
[PMCommand/profile] Part 1.1: Do some tweaking!
And uh. Break time

Commit:
[IPC] Remove unused code

Commit:
[Web] /dashboard/management/logs
A completely new maintainer console page! Displaying the most recent 10 console log lines, and streaming any new ones instantly.
Because the logs are streamed from the console, the verbosity can be modified by modifying the consoleLevel value in config.js.
Currently doesn't display any colors, maybe should be added in the future? If we get around to it; sure.

Commit:
[Ready] Temp fix a Ready bug
The server-config removal collects all guild ID's  from all shards using IPC, however, if a shard has not been fully initialized yet, it will return an empty array, thus leaving out servers whose config would then be destroyed. Perhaps not do this on startup, but instead make it something manual either the server's owner or a maintainer can do?

Commit:
[Web] /dashboard/management/eval
An evaluation page, where you can evaluate bits of JavaScript code within the Maintainer Console on either the master, a specific Shard, or all shards. Requires the "eval" permission to access!

Commit:
[IPC] Fix JSON parsing error

Commit:
[Dependencies] Updoots!

Commit:
[SharedCommands/Events] Shared command system :tada:

Shared commands are the most basic commands. They don't expect much info.
It's at the developers choice whether to make them a separate command (like eval) or make a public and private command which in turn, will run the shared command (and not defining it in commands.js in that case)
Things are still experimental, but they work!

Commit:
[*] Fix Gilbert's observations

Commit:
[GAB.MaintainerMessageHandler] Fix for canRunSharedCommand check

Commit:
[Events] Remvoe reference to Database and fix bug

Commit:
[Web] Blog fix

Commit:
[*] GLOBALLLLSSSS

Commit:
[Web] Fixes

Commit:
[MessageCreate] Fix bug

Commit:
[Web] configJSON Update fix

Commit:
[Schema] Changes to strikes, that will allow good things.
Note that this will not affect old strikes.

Commit:
[*] Fix mongoose $pushAll, make client.getCommandPrefix a non-promise
Make the public AFK command, make sure to get the command only if a string is provided
Add Client.getPublicCommandName, same as Client.getSharedCommandName

Commit:
[Mongoose] Due to mongoose updates, make sure to not use $pushAll..
Damn it why must u break things!

Commit:
[Fixes] Fix some bugs / New Bot added embed / Fix adding of roles if bot doesn't have perms

Commit:
[*] Add a standard Constants file
This file will host all our constants. For ease of purpoise!

Commit:
[Encryption] New Encryption Module!
Now, you may think: why? What do we need this for?
We have some sick things planned that involve message edits..
But we are forced to encrypt and save the data encrypted in the database, and decrypt it when needed.
As such, we've implemented this. It'll get used later on

Commit:
String updates

Commit:
More string updates

(For the sake of consistency, only use the verb "to use" when instructing a user to run a command!)

Commit:
Even more string updates

Commit:
[Encryption] Fix some small  bugs

Commit:
[Events/Message] String Update + Massive bug fix
Since we support aliases, command usage would save the usage depending on what command you entered
So, for instance: `ping` may have an alias of `pong`. It would save `ping` if you run `ping` and `pong` if you ran `pong`
This fixes that!

Commit:
Typo

Commit:
[Commands] Make reload a shared command!
Also some web. Sorry Gilberto

Commit:
Wrong Check for reload

Commit:
Nleh, mistype

Commit:
Even more "even more string updates"

Commit:
Some more string updates

Commit:
[IPC] Fix slight JSON error

Commit:
[Bugs] Fix random not being a function

Commit:
[Events/MessageCreate] Fix undefined Constants, alias bugs, and random not being a function

Commit:
[Commands/About] About command 97% done

Commit:
Part 2 of the about command

Commit:
[Commands/alert] Alert command is done

Commit:
[*] Change line type from CRLF to LF

Commit:
[Web/Boot/Discord] /dashboard/management/shards
Shut GAwesomeBot down, restart GAwesomeBot!
Also, freeze, reset or restart any shard!

Commit:
[Bugs/Client] Add client to commands, and fix eval gist

Commit:
[Dependencies] Another update

Commit:
[Discord.js] Fix undefined

Commit:
[Base Commands] Add constants to the docs

Commit:
[Discord] Improvement, I guess

Commit:
[PMCommands] Fix bug gilbert found

Commit:
[WIP] Please ignroe

Commit:
[Web] Ooh, what is this?
Gilbert cleaning his working tree.

Commit:
Update maintainer perm defaults

With the evaluate permission, maintainers can "elevate" their perms to any level. Free execution of JavaScript enables users to have full control over the entirety of GAB, so it should be defaulted to The Host only. The shutdown commands are "shortcuts" of eval commands that sudo maintainers can have access to without undermining the permission system.

Commit:
[Web] Some whitespace improvements
And string updates!

Commit:
[Web] Woops
Turns out the blog/wiki composer/editor was always broken!

Commit:
[Web] Even more whitespacing updates

Commit:
[Github] Template updates

Commit:
[Cache] Fix bugs that Gilbert found
They're fixed! So thats good!

Commit:
[Commands] about command updates

Commit:
[Commands] Even more command updates

Commit:
[Discord] Oof, bug fixes!

Commit:
[Sharder] Mode option for the future

Commit:
[Web] Fix /userlist bug

Commit:
[Discord] They call me Gilbert, Dirty Gilbert

Commit:
[*] Clean Dirty Gilbert's.. Dirt

Commit:
[Structures] Fix my dumb

Commit:
[Discord] They call me Gilbert, Dirty Gilbert

Commit:
[Discord.js] Saw nothing

Commit:
[Trivia] Just me nitpicking.

Commit:
[Trivia] More nitpicks

This ones on me, I promise!

Commit:
[Eval] Fix some dumbs

Commit:
[Dependencies]

Commit:
[*] Move these

Commit:
[Cache/*] Fix cache issues

Commit:
[*] Gilbert shush

Commit:
Woopsie daisy you did NOT see this

Commit:
[DB] Shh

Commit:
[*] Renamed the maintainer message handler to shared command handler
Renamed the property configJSON for shared commands to "perm"
Added "any" to represent that anyone can run the command
Allow the GAB HOST to run commmands that are set for the host only

Commit:
[Commands] Moved Colors to the Constants file

Commit:
[Dependencies]

Commit:
[Commands/Utils] Reaction Menu and anime commands are done!

Commit:
[Configs] Update permission bits for the invite URL

This will ensure that the bot works at its maximum capacity!

Commit:
[Discord.js] Less duplicated code

Commit:
[Anime] Clean it up a bit

Commit:
[Anime] Nitpiks

Commit:
[*] Poll command, Relay Commands, so much more!
Now now, calm down Vlad. Remember this is WIP code, it's a draft. If you don't like the way Relay commands work, you can refactor them to your liking.

The Polls Public command isn't present yet either, I'm not a machine geez.

Commit:
[Deps]

Commit:
[Commands] More work!

Commit:
Woops, dun goofed!

Commit:
Spacing

Commit:
[Deps] Ignore yarn being dum

Commit:
Update PULL_REQUEST_TEMPLATE.md

Commit:
[Evens/Message-SpamHandler] Fix a quite funny bug
Let the bot ignore bots from the spam handler.
http://wait-it.works.wtf/vlag/2017/12/69L896.png

Commit:
[Commands/Avatar] Done did avatar

Commit:
[Commands/Web] The Public Poll command
And some misc. fixes

Commit:
[Updater] New ConfigWarning message
Moved the version/branch not found warning to configWarnings

Commit:
[Public] ???

Commit:
[ServerSchema] Add ban gif property
Will be used in the upcomming ban command :eyes:

Commit:
[Yarn] Ignore me

Commit:
[Utils] Fix isURL

Commit:
[*] Ban command! Some work on modlogs!
Disabled fastboot (do we need it anyways?)
Revome duplicated winston instance
New Color
New Error Messages
And more blobs!

Commit:
Fix this

Commit:
Color change!

Commit:
[Message] Fix a bug where @bot help wouldn't respond if the chatterbot was off

Commit:
Aliases!

Commit:
[Ban] Make ban also make a modlog entry

Commit:
[Modlog] Fix some bugs

Commit:
[Web] Change the way we sort channels for things that require channels!

Now now Gilbert, before you scream at me for touching web, let me explain.

Basically, channel.position is the calculated position based on the category (or in some cases, no category)
Raw position is the actual raw position, from 0 to w/e is the last channel
In ourcase, we want to sort by raw position
This makes the channel list look almost identical to Discords one

Commit:
[Commands.js] Slight tweaks, more to come!

Commit:
[Web] More minor tweaks ala Vlag

Commit:
[Ban] Fix sum bugs!

Commit:
[Avatar] Minor tweaks

Commit:
Avatar cmd but better!

Commit:
QoL

Commit:
[*] Fix some bugs and make afk better!

Commit:
QoL

Commit:
[*] Fix some dumbs I made!

Commit:
[Commands] giveaway!

Commit:
Woops, I did a booboo

Commit:
🤪

Commit:
[MessageCreate] Fix goof

Commit:
[Calc] Calc command, done right!

Commit:
[Ban] tf you write something and find a bug

Commit:
[Calc] You take a ss and see that it would look better as no title
Bleh

Commit:
[Cat] Meow!

Commit:
[Commands] Servernick!

Commit:
[Commands] Calc, done worse

Calc is pretty broken on my instance, Vladdie save us all!
Also, text constants for development consistency whoo hoo

Commit:
[Commands] Color constants

Commit:
[Modules] Add random dogs in the mix!

Commit:
[Calc] But better! But still, Gilbert has to touch it!
asically, you can now re-reload the command! :tada:
Plus, the new ChildProcessManager will come in handy for extensions
Oh yeah! Calc may be slower now. Gilbert may fix it. Kek
And some color name changes.

Commit:
[*] Consistency
And I do mean,
C 👏 O 👏 N 👏 S 👏 I 👏S 👏 T 👏 E 👏 N 👏 C 👏 Y

Commit:
QoL

Commit:
[Ban] Mention the member if they're in the guild, in case theres multiple people you can see who you're actually banning
Also message expires in 2 mins

Commit:
[*] Renames! From ChildProcessManagers and such to Workers

Commit:
[Workers] Make the worker be one for each shard
Weoffload certain tasks to it to aleviate the main bot
Things like extensions will require this
Plus, this opens up a lot of free room to use things that we don't trust 100%
(see: Extensions, or whatnot)

Commit:
Update CODEOWNERS

Commit:
[Commands/Web/Discord] giveaway
+ logMessage fixes

Commit:
[Web] Added reactivity to name-display

Commit:
[catfact]!! Paginated menus! Fixes! Eggs?
Moved Emojis to Constants
And god bless rebasing won't suck!

Commit:
[Misc] Crazy idea: Move all API URLs to constants
Because then we can change em on the fly. Ish

Commit:
Typo

Commit:
[choose] Choose command done

Commit:
[Misc] Consistency

Commit:
[Commands] Make Trivia more consistent

Commit:
[Eslint] Uh, new line anyone?

Commit:
[Lottery] Consistent!

Commit:
[Lottery] Fix bug

Commit:
[Web] Activity page bans!
And some optimizations. To ban a guild from the activity page, simply navigate to /activity/servers, find the guild you want to ban, and click on the cross next to their name.

Commit:
[ConfigJSON] Updoots

SunburntRock89 commented:
This pains me [](http://gawesomebotl.works.wtf/how/5d244.png)

Gilbert142 committed:
Public Data unbans & Socket.io fixes

Gilbert142 committed:
[Web] Ban gifs/customization!

SunburntRock89 commented:
I regret ever commenting because now I'm subscribed lol

wasdennnoch commented:
I mean, you can just unsubscribe via the button in the right sidebar. Or unwatch the project at the top of the page.

Commit:
[travis] Shut up

Commit:
[ObjectDefines] Add trimAll to all arrays
Basically takes all strings and trims em

Gilbert142 committed:
[Web] Save button updates

Gilbert142 committed:
[Web] Default buttons

Commit:
[Worker] Make winston global

Commit:
[Web] Help button update

Commit:
[Modules] Out with this!

Commit:
[Dependencies] Updates!

Commit:
[Commands] remindme

Commit:
[Commands] say whaaat?

Commit:
[Mongoose] Prepare for mongoose v5 and fix some bugs
And bumb dependencies

Commit:
[Poll] Touched some embeds and done some Vlag-Approved optimization

Commit:
[Commands] Tags! Done not-badly.

Commit:
[Commands] !?!?!?

Commit:
[Utils] Markdown tables!

Commit:
[tags] Fix some misc weird things.
The text really loved sticking to itself. Now it listens like a good boi

Commit:
Misc

Commit:
[Commands/Discord] ranks!
Made the ranks command, cleaned up checkRank, fixed a logMessage bug.

Commit:
[Discord] pfft

Commit:
[PostData] Remove the DiscordBotOrg code, will implement using my module

Commit:
[Discord] pfft

Commit:
NootNoot we've got new dependency updates boi

Commit:
[convert] Say waa?
Convert your currency or your length...if that's what you're into..
You'll need an Open Exchange Rate Key (which is free, for 1000 queries per month)
We also cache the results in a local JSON, so while the results at boot may be outdated (they will update over time), you won't go nuts on your usage quota!

Commit:
[Misc] Fix copy paste
Calculating conversions. :lul:

Commit:
[commands.js] Small tweaks

Commit:
[package.json] Improve the warning message for windows users only.
Now with more colors! Awooo

Commit:
[Travis] Fix travis for ReactionBasedMenu

Commit:
[convert] Fix dumbs

Commit:
Convert rounds up cash

Commit:
[Events] Use global value instead of this.configJSON
Because it doesn't update properly. @KingDGrizzle update the rest please!

Commit:
[Constants] Woopsies doosels

Commit:
[Commands] Count, and debug fixes

vitalkanev commented:
I guess done @GilbertGobbels?

KingDGrizzle commented:
Please note that this can be counted as `PR 1 out of many` when it comes to the rewrite. Expect more PRs

TTtie commented:
that reminds me of the thing that I should work on #202
